src/Microsoft.AspNet.OData.Shared/Builder/ODataConventionModelBuilder.cs

```diff
@@ -1080,21 +1080,48 @@ private void ApplyPropertyConvention(IEdmPropertyConvention propertyConvention,
             }
         }
 
-        private static Dictionary<Type, List<Type>> BuildDerivedTypesMapping(IWebApiAssembliesResolver assemblyResolver)
+        private static Dictionary<Type, Type[]> BuildDerivedTypesMapping(IWebApiAssembliesResolver assemblyResolver)
         {
-            IEnumerable<Type> allTypes = TypeHelper.GetLoadedTypes(assemblyResolver).Where(t => TypeHelper.IsVisible(t) && TypeHelper.IsClass(t) && t != typeof(object));
-            Dictionary<Type, List<Type>> allTypeMapping = allTypes.Distinct().ToDictionary(k => k, k => new List<Type>());
-
-            foreach (Type type in allTypes)
+            Predicate<Type> verifyType = t => TypeHelper.IsVisible(t) && TypeHelper.IsClass(t) && t != typeof(object);
```

Comment:
>TypeHelper.IsVisible(t) [](start = 46, length = 23)

Note: this is already checked in GetLoadedTypes -- don't need to do it again here.  Can any of these other checks be pushed to GetLoadedTypes? i.e., don't we awlays want to exclued typeof(object)?

Reply:
Yes, IsVisible can remove. 

TypeHelper.GetLoadedTypes, as its name, is to loaded types, doesn't care about it's class or object. So, let's keep the method atomic.

---
In reply to: [598862955](https://github.com/OData/WebApi/pull/2436#discussion_r598862955) [](ancestors = 598862955)

Reply:
"verifyType" action is also used to verify the base type. 
BaseType is loaded at Line 1096, that's not from "GetLoadedTypes".

So, all these checks are needed here.

---
In reply to: [598862955](https://github.com/OData/WebApi/pull/2436#discussion_r598862955) [](ancestors = 598862955)